File: String/Palindrome_Partition_2.py
"""
Given a string s, partition s such that every substring of the partition is a palindrome.
Return all possible palindrome partitioning of s.

Dynamic programming approach.

Time complexity: O(N*2^N), where N is the length of the string S.
"""

class Solution:
    def partition(self,s:str):
        length = len(s)
        currentList = []
        dp = [[False for i in range(length)] for j in range(length)]
        result = []

        def dfs(start):
            if(start >= length):
                result.append(currentList.copy())
            for end in range(start,length):
                if(s[start]==s[end] and (end-start <=2 or dp[start+1][end-1])):
                    dp[start][end]=True
                    currentList.append(s[start:end+1])
                    dfs(end+1)
                    currentList.pop()

        dfs(0)
        return result

if __name__=="__main__":
    s = input()
    c = Solution()
    for x in c.partition(s):
        print(x, end=" ")

File: String/Palindrome_Partition_2.cpp
/*
Given a string s, partition s such that every substring of the partition is a palindrome.
Return all possible palindrome partitioning of s.

Dynamic programming approach.

Time complexity: O(N*2^N), where N is the length of the string S.
*/


#include <bits/stdc++.h>
using namespace std;

vector<vector<string>> result;
vector<string> currentList;

class Solution {
    public:
    static vector<vector<string>> partition(string s) {
        int len = s.length();
        vector<vector<bool>> dp(len, vector<bool> (len,false));
        dfs(0,s,dp);
        return result;
    }
    private:
    static void dfs(int start,string &s,vector<vector<bool>> &dp) {
        if(start >= s.length()) {
            result.emplace_back(currentList);
        }
        for(auto end = start; end < s.length(); ++end) {
            if(s[start]==s[end] && (end-start <= 2 || dp[start+1][end-1])) {
                dp[start][end] = true;
                currentList.emplace_back(s.substr(start,end-start+1));
                dfs(end+1,s,dp);
                currentList.pop_back();
            }
        }
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    string s = "aab";
    for(auto &x : Solution::partition(s)) {
        for(auto &y : x) {
            cout << y << " ";
        }
    }
    return 0;
}
